File: phases/04-algorithms-and-complexity/14-graph-algorithms-ii-dijkstra-bellman-ford-a/code/main.py
"""
Graph Algorithms II — Dijkstra, Bellman-Ford, A*
Phase 04 — Algorithms & Complexity Analysis

Shortest path algorithms with path reconstruction and benchmarks.
"""

import heapq
import math
import random
import time
from typing import Optional


# ---------------------------------------------------------------------------
# Dijkstra
# ---------------------------------------------------------------------------

def dijkstra(graph: dict[int, list[tuple[int, float]]], src: int):
    """Single-source shortest paths (non-negative weights).
    Returns (dist, prev) dicts."""
    dist = {v: float('inf') for v in graph}
    prev = {v: None for v in graph}
    dist[src] = 0
    pq = [(0, src)]

    while pq:
        d, u = heapq.heappop(pq)
        if d > dist[u]:
            continue
        for v, w in graph[u]:
            nd = d + w
            if nd < dist[v]:
                dist[v] = nd
                prev[v] = u
                heapq.heappush(pq, (nd, v))

    return dist, prev


def reconstruct_path(prev: dict, target: int) -> list[int]:
    path = []
    node = target
    while node is not None:
        path.append(node)
        node = prev[node]
    path.reverse()
    return path


# ---------------------------------------------------------------------------
# Bellman-Ford
# ---------------------------------------------------------------------------

def bellman_ford(edges: list[tuple[int, int, float]], V: int, src: int):
    """Single-source shortest paths (arbitrary weights).
    Returns (dist, prev, negative_cycle). negative_cycle is None or a list
    of vertices forming a negative cycle."""
    dist = [float('inf')] * V
    prev_arr = [None] * V
    dist[src] = 0

    for _ in range(V - 1):
        updated = False
        for u, v, w in edges:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                prev_arr[v] = u
                updated = True
        if not updated:
            break

    # Negative cycle detection
    for u, v, w in edges:
        if dist[u] + w < dist[v]:
            cycle_node = v
            for _ in range(V):
                cycle_node = prev_arr[cycle_node]
            cycle = []
            node = cycle_node
            while True:
                cycle.append(node)
                node = prev_arr[node]
                if node == cycle_node:
                    cycle.append(node)
                    break
            cycle.reverse()
            return dist, prev_arr, cycle

    return dist, prev_arr, None


def reconstruct_path_bf(prev_arr: list, target: int) -> list[int]:
    path = []
    node = target
    while node is not None:
        path.append(node)
        node = prev_arr[node]
    path.reverse()
    return path


# ---------------------------------------------------------------------------
# A* Search
# ---------------------------------------------------------------------------

def astar(grid: list[list[int]], src: tuple, goal: tuple):
    """A* on a 2D grid. 0=walkable, 1=obstacle.
    Returns (path, cost) or (None, inf) if unreachable."""
    rows, cols = len(grid), len(grid[0])

    def h(pos):
        return abs(pos[0] - goal[0]) + abs(pos[1] - goal[1])

    def neighbors(pos):
        r, c = pos
        for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
            nr, nc = r + dr, c + dc
            if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == 0:
                yield (nr, nc)

    g_score = {src: 0}
    prev = {src: None}
    pq = [(h(src), 0, src)]

    while pq:
        f, g, u = heapq.heappop(pq)
        if u == goal:
            path = []
            node = goal
            while node is not None:
                path.append(node)
                node = prev[node]
            path.reverse()
            return path, g

        if g > g_score.get(u, float('inf')):
            continue

        for v in neighbors(u):
            ng = g + 1
            if ng < g_score.get(v, float('inf')):
                g_score[v] = ng
                prev[v] = u
                heapq.heappush(pq, (ng + h(v), ng, v))

    return None, float('inf')


# ---------------------------------------------------------------------------
# Comparison Benchmarks
# ---------------------------------------------------------------------------

def generate_random_graph(V: int, E: int, max_weight: int = 100) -> dict:
    graph = {i: [] for i in range(V)}
    edges = []
    for _ in range(E):
        u = random.randint(0, V - 1)
        v = random.randint(0, V - 1)
        w = random.randint(1, max_weight)
        graph[u].append((v, w))
        edges.append((u, v, w))
    return graph, edges


def generate_weighted_grid(rows: int, cols: int, obstacle_pct: float = 0.2):
    grid = [[0] * cols for _ in range(rows)]
    for r in range(rows):
        for c in range(cols):
            if random.random() < obstacle_pct:
                grid[r][c] = 1
    return grid


def benchmark():
    print("=== Benchmark: Dijkstra vs Bellman-Ford ===")
    print(f"{'V':>6} {'E':>8} {'Dijkstra (ms)':>14} {'Bellman-Ford (ms)':>18}")
    print("-" * 50)

    for V in [100, 500, 1000]:
        E = V * 4
        graph, edges = generate_random_graph(V, E)

        t0 = time.perf_counter()
        for _ in range(5):
            dijkstra(graph, 0)
        dt = (time.perf_counter() - t0) / 5 * 1000

        t0 = time.perf_counter()
        for _ in range(5):
            bellman_ford(edges, V, 0)
        bt = (time.perf_counter() - t0) / 5 * 1000

        print(f"{V:>6} {E:>8} {dt:>14.2f} {bt:>18.2f}")


# ---------------------------------------------------------------------------
# Main — demonstrations
# ---------------------------------------------------------------------------

def main() -> None:
    # --- Dijkstra demo ---
    print("=" * 60)
    print("DIJKSTRA'S ALGORITHM")
    print("=" * 60)

    graph = {
        0: [(1, 4), (2, 1)],
        1: [(3, 1)],
        2: [(1, 2), (3, 5)],
        3: [(4, 3)],
        4: [],
    }
    dist, prev = dijkstra(graph, 0)
    print(f"Shortest distances from 0: {dist}")
    print(f"Path 0 -> 4: {reconstruct_path(prev, 4)}")
    print(f"Path 0 -> 1: {reconstruct_path(prev, 1)}")

    # --- Bellman-Ford demo ---
    print()
    print("=" * 60)
    print("BELLMAN-FORD ALGORITHM")
    print("=" * 60)

    edges = [
        (0, 1, 4), (0, 2, 1),
        (1, 3, 1), (2, 1, 2),
        (2, 3, 5), (3, 4, 3),
    ]
    dist_bf, prev_bf, cycle = bellman_ford(edges, 5, 0)
    print(f"Shortest distances from 0: {dist_bf}")
    print(f"Path 0 -> 4: {reconstruct_path_bf(prev_bf, 4)}")

    # Negative cycle demo
    neg_edges = [
        (0, 1, 1), (1, 2, -3),
        (2, 3, -1), (3, 1, 2),
    ]
    _, _, cycle = bellman_ford(neg_edges, 4, 0)
    print(f"Negative cycle detected: {cycle}")

    # No negative cycle
    safe_edges = [
        (0, 1, 1), (1, 2, 2),
        (2, 3, 3),
    ]
    _, _, cycle = bellman_ford(safe_edges, 4, 0)
    print(f"Negative cycle (should be None): {cycle}")

    # --- A* demo ---
    print()
    print("=" * 60)
    print("A* SEARCH")
    print("=" * 60)

    grid = [
        [0, 0, 0, 0, 0],
        [0, 1, 1, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 1, 1, 0],
        [0, 0, 0, 0, 0],
    ]
    path, cost = astar(grid, (0, 0), (4, 4))
    print(f"Path from (0,0) to (4,4): {path}")
    print(f"Cost: {cost}")

    # Unreachable
    blocked = [
        [0, 1, 0],
        [1, 1, 1],
        [0, 1, 0],
    ]
    path, cost = astar(blocked, (0, 0), (2, 2))
    print(f"Blocked path: {path}, cost: {cost}")

    # --- Correctness checks ---
    print()
    print("=" * 60)
    print("CORRECTNESS CHECKS")
    print("=" * 60)

    # Dijkstra vs Bellman-Ford on non-negative graph
    g, e = generate_random_graph(50, 200, 100)
    d_dij, _ = dijkstra(g, 0)
    d_bf, _, _ = bellman_ford(e, 50, 0)
    match = all(
        abs(d_dij.get(v, float('inf')) - (d_bf[v] if v < 50 else float('inf'))) < 1e-9
        for v in g
    )
    print(f"Dijkstra == Bellman-Ford on non-negative graph: {match}")

    # A* returns optimal path
    small_grid = [[0] * 10 for _ in range(10)]
    path_astar, cost_astar = astar(small_grid, (0, 0), (9, 9))
    print(f"A* on empty 10x10 grid: cost={cost_astar} (expected 18)")

    # --- Benchmark ---
    print()
    benchmark()

    print()
    print("All demos complete.")


if __name__ == "__main__":
    main()

File: phases/04-algorithms-and-complexity/14-graph-algorithms-ii-dijkstra-bellman-ford-a/code/main.rs
//! Graph Algorithms II — Dijkstra, Bellman-Ford, A*
//! Phase 04 — Algorithms & Complexity Analysis
//!
//! Dijkstra and Bellman-Ford with path reconstruction using BinaryHeap.

use std::cmp::Ordering;
use std::collections::BinaryHeap;

// ---------------------------------------------------------------------------
// Dijkstra
// ---------------------------------------------------------------------------

#[derive(Eq, PartialEq)]
struct State {
    cost: i64,
    node: usize,
}

impl Ord for State {
    fn cmp(&self, other: &Self) -> Ordering {
        other.cost.cmp(&self.cost) // min-heap via reversed ordering
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

pub fn dijkstra(adj: &[Vec<(usize, i64)>], src: usize) -> (Vec<i64>, Vec<Option<usize>>) {
    let n = adj.len();
    let mut dist = vec![i64::MAX; n];
    let mut prev = vec![None; n];
    dist[src] = 0;

    let mut pq = BinaryHeap::new();
    pq.push(State { cost: 0, node: src });

    while let Some(State { cost, node }) = pq.pop() {
        if cost > dist[node] {
            continue;
        }
        for &(v, w) in &adj[node] {
            let nd = cost + w;
            if nd < dist[v] {
                dist[v] = nd;
                prev[v] = Some(node);
                pq.push(State { cost: nd, node: v });
            }
        }
    }

    (dist, prev)
}

pub fn reconstruct_path(prev: &[Option<usize>], target: usize) -> Vec<usize> {
    let mut path = Vec::new();
    let mut cur = Some(target);
    while let Some(node) = cur {
        path.push(node);
        cur = prev[node];
    }
    path.reverse();
    path
}

// ---------------------------------------------------------------------------
// Bellman-Ford
// ---------------------------------------------------------------------------

pub fn bellman_ford(
    edges: &[(usize, usize, i64)],
    v: usize,
    src: usize,
) -> (Vec<i64>, Vec<Option<usize>>, Option<Vec<usize>>) {
    let mut dist = vec![i64::MAX; v];
    let mut prev = vec![None; v];
    dist[src] = 0;

    for _ in 0..(v - 1) {
        let mut updated = false;
        for &(u, vtx, w) in edges {
            if dist[u] != i64::MAX && dist[u] + w < dist[vtx] {
                dist[vtx] = dist[u] + w;
                prev[vtx] = Some(u);
                updated = true;
            }
        }
        if !updated {
            break;
        }
    }

    // Negative cycle detection
    for &(u, vtx, w) in edges {
        if dist[u] != i64::MAX && dist[u] + w < dist[vtx] {
            let mut cycle_node = vtx;
            for _ in 0..v {
                cycle_node = prev[cycle_node].unwrap_or(0);
            }
            let mut cycle = Vec::new();
            let mut node = cycle_node;
            loop {
                cycle.push(node);
                node = prev[node].unwrap_or(0);
                if node == cycle_node {
                    cycle.push(node);
                    break;
                }
            }
            cycle.reverse();
            return (dist, prev, Some(cycle));
        }
    }

    (dist, prev, None)
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

fn main() {
    // --- Dijkstra demo ---
    println!("{}", "=".repeat(60));
    println!("DIJKSTRA'S ALGORITHM");
    println!("{}", "=".repeat(60));

    let adj = vec![
        vec![(1, 4), (2, 1)],
        vec![(3, 1)],
        vec![(1, 2), (3, 5)],
        vec![(4, 3)],
        vec![],
    ];

    let (dist, prev) = dijkstra(&adj, 0);
    println!("Shortest distances from 0: {:?}", dist);
    let path = reconstruct_path(&prev, 4);
    println!("Path 0 -> 4: {:?}", path);
    let path1 = reconstruct_path(&prev, 1);
    println!("Path 0 -> 1: {:?}", path1);

    // --- Bellman-Ford demo ---
    println!();
    println!("{}", "=".repeat(60));
    println!("BELLMAN-FORD ALGORITHM");
    println!("{}", "=".repeat(60));

    let edges = vec![
        (0, 1, 4), (0, 2, 1),
        (1, 3, 1), (2, 1, 2),
        (2, 3, 5), (3, 4, 3),
    ];
    let (dist_bf, prev_bf, cycle) = bellman_ford(&edges, 5, 0);
    println!("Shortest distances from 0: {:?}", dist_bf);
    let path_bf = reconstruct_path(&prev_bf, 4);
    println!("Path 0 -> 4: {:?}", path_bf);

    // Negative cycle demo
    let neg_edges = vec![
        (0, 1, 1), (1, 2, -3),
        (2, 3, -1), (3, 1, 2),
    ];
    let (_, _, cycle) = bellman_ford(&neg_edges, 4, 0);
    println!("Negative cycle detected: {:?}", cycle);

    // No negative cycle
    let safe_edges = vec![
        (0, 1, 1), (1, 2, 2),
        (2, 3, 3),
    ];
    let (_, _, cycle) = bellman_ford(&safe_edges, 4, 0);
    println!("Negative cycle (should be None): {:?}", cycle);

    // --- Correctness checks ---
    println!();
    println!("{}", "=".repeat(60));
    println!("CORRECTNESS CHECKS");
    println!("{}", "=".repeat(60));

    // Verify Dijkstra == Bellman-Ford on non-negative graph
    let v = 50;
    let test_edges = vec![
        (0, 1, 4), (0, 2, 1), (1, 3, 1), (2, 1, 2),
        (2, 3, 5), (3, 4, 3), (4, 5, 2), (1, 5, 10),
    ];
    let mut test_adj = vec![vec![]; v];
    for &(u, vtx, w) in &test_edges {
        test_adj[u].push((vtx, w));
    }
    let (d_dij, _) = dijkstra(&test_adj, 0);
    let (d_bf, _, _) = bellman_ford(&test_edges, v, 0);
    let match_ok = d_dij.iter().zip(d_bf.iter()).all(|(a, b)| a == b);
    println!("Dijkstra == Bellman-Ford on non-negative: {}", match_ok);

    println!();
    println!("All demos complete.");
    println!("\nRun `cargo test` to verify correctness.");
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dijkstra_simple() {
        let adj = vec![
            vec![(1, 4), (2, 1)],
            vec![(3, 1)],
            vec![(1, 2), (3, 5)],
            vec![(4, 3)],
            vec![],
        ];
        let (dist, prev) = dijkstra(&adj, 0);
        assert_eq!(dist[0], 0);
        assert_eq!(dist[1], 3);
        assert_eq!(dist[2], 1);
        assert_eq!(dist[3], 4);
        assert_eq!(dist[4], 7);

        let path = reconstruct_path(&prev, 4);
        assert_eq!(path, vec![0, 1, 3, 4]);
    }

    #[test]
    fn dijkstra_disconnected() {
        let adj = vec![vec![(1, 5)], vec![], vec![]];
        let (dist, _) = dijkstra(&adj, 0);
        assert_eq!(dist[0], 0);
        assert_eq!(dist[1], 5);
        assert_eq!(dist[2], i64::MAX);
    }

    #[test]
    fn bellman_ford_no_negative() {
        let edges = vec![
            (0, 1, 4), (0, 2, 1),
            (1, 3, 1), (2, 1, 2),
            (2, 3, 5), (3, 4, 3),
        ];
        let (dist, prev, cycle) = bellman_ford(&edges, 5, 0);
        assert!(cycle.is_none());
        assert_eq!(dist[4], 8);

        let path = reconstruct_path(&prev, 4);
        assert_eq!(path, vec![0, 1, 3, 4]);
    }

    #[test]
    fn bellman_ford_negative_cycle() {
        let edges = vec![
            (0, 1, 1), (1, 2, -3),
            (2, 3, -1), (3, 1, 2),
        ];
        let (_, _, cycle) = bellman_ford(&edges, 4, 0);
        assert!(cycle.is_some());
        let c = cycle.unwrap();
        assert!(c.len() >= 2);
    }

    #[test]
    fn bellman_ford_no_cycle_safe() {
        let edges = vec![
            (0, 1, 1), (1, 2, 2),
            (2, 3, 3),
        ];
        let (_, _, cycle) = bellman_ford(&edges, 4, 0);
        assert!(cycle.is_none());
    }

    #[test]
    fn dijkstra_matches_bellman_ford() {
        let edges = vec![
            (0, 1, 4), (0, 2, 1), (1, 3, 1), (2, 1, 2),
            (2, 3, 5), (3, 4, 3), (4, 5, 2), (1, 5, 10),
        ];
        let mut adj = vec![vec![]; 6];
        for &(u, v, w) in &edges {
            adj[u].push((v, w));
        }
        let (d_dij, _) = dijkstra(&adj, 0);
        let (d_bf, _, cycle) = bellman_ford(&edges, 6, 0);
        assert!(cycle.is_none());
        for i in 0..6 {
            assert_eq!(d_dij[i], d_bf[i], "mismatch at vertex {}", i);
        }
    }
}
